From b6bb213e59e5d52f4d32f562351a4c0422ed74ef Mon Sep 17 00:00:00 2001
From: Li Jun <jun.li@nxp.com>
Date: Fri, 27 Oct 2017 02:28:15 +0800
Subject: [PATCH 2724/5242] MLK-16697 staging: typec: tcpci: add notfication
 for device attach

commit  ada8b67b8a966439cf38a68be6618b4d6c22dc3f from
https://source.codeaurora.org/external/imx/linux-imx.git

Some usb device driver can't know the connect and disconnect to host
if the vbus is always on, if use typec we can rely on cc line status
to know that, so add a notification to let controller driver know
device attach and detach from host.

Acked-by: Peter Chen <peter.chen@nxp.com>
Signed-off-by: Li Jun <jun.li@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/staging/typec/tcpci.c |   20 ++++++++++++++++++--
 1 file changed, 18 insertions(+), 2 deletions(-)

diff --git a/drivers/staging/typec/tcpci.c b/drivers/staging/typec/tcpci.c
index e1a566c..7903313 100644
--- a/drivers/staging/typec/tcpci.c
+++ b/drivers/staging/typec/tcpci.c
@@ -30,6 +30,7 @@ struct tcpci {
 	struct regmap *regmap;
 
 	bool controls_vbus;
+	bool attached;
 	struct gpio_desc *ss_sel_gpio;
 
 	struct tcpc_dev tcpc;
@@ -44,6 +45,7 @@ struct tcpci_chip {
 
 static const unsigned int tcpci_extcon_cable[] = {
 	EXTCON_USB_HOST,
+	EXTCON_USB,
 	EXTCON_NONE,
 };
 
@@ -261,6 +263,11 @@ static int tcpci_get_cc(struct tcpc_dev *tcpc,
 				 TCPC_CC_STATUS_CC2_MASK,
 				 reg & TCPC_CC_STATUS_TERM);
 
+	if ((*cc1 == TYPEC_CC_OPEN) && (*cc2 == TYPEC_CC_OPEN))
+		tcpci->attached = false;
+	else
+		tcpci->attached = true;
+
 	return 0;
 }
 
@@ -329,10 +336,19 @@ static int tcpci_set_roles(struct tcpc_dev *tcpc, bool attached,
 	if (ret < 0)
 		return ret;
 
-	if (data == TYPEC_HOST)
+	if (data == TYPEC_HOST) {
+		extcon_set_state_sync(tcpci->edev, EXTCON_USB, false);
 		extcon_set_state_sync(tcpci->edev, EXTCON_USB_HOST, true);
-	else
+	} else {
 		extcon_set_state_sync(tcpci->edev, EXTCON_USB_HOST, false);
+		/*
+		 * Instead of use 'attached' input, we need
+		 * the real HW connection status to notify
+		 * USB device controller driver the attach
+		 * and dettach event to host.
+		 */
+		extcon_set_state_sync(tcpci->edev, EXTCON_USB, tcpci->attached);
+	}
 
 	return 0;
 }
-- 
1.7.9.5

